File: internal/maintenanceservice/memory.go
// memory.go hosts the Week 28 memory-manager methods (per-project memory/ +
// agent-memory/) on MaintenanceService — kept off service.go, matching
// agents.go/skills.go/instructions.go. MaintenanceService is the right home: it
// already owns the SSH gate, s.mu, and the trash engine DeleteMemoryFile needs,
// and every method resolves the same EffectivePath root so reads/writes/deletes
// stay on one tree for a custom-root user. Dirs are addressed by kind-prefixed
// ID (never a client path), so the files layer's deterministic resolver is the
// only place a path is derived.
package maintenanceservice

import (
	"claude-devtools/internal/config"
	"claude-devtools/internal/files"
	"claude-devtools/internal/maintenance"
)

// ListMemoryDirs enumerates the addressable memory dirs (per-project + per-agent)
// with kind-prefixed IDs. Read-only: no SSH gate.
func (s *MaintenanceService) ListMemoryDirs() ([]files.MemoryDir, error) {
	root := s.config.GetClaudeRootInfo().EffectivePath
	return files.ListMemoryDirs(root)
}

// MemoryIntegrity returns one dir's fact files + the four integrity finding
// kinds (orphan/dangling-index/dangling-link/duplicate-slug). Read-only: no SSH
// gate, no filesystem writes.
func (s *MaintenanceService) MemoryIntegrity(dirID string) (files.MemoryReport, error) {
	root := s.config.GetClaudeRootInfo().EffectivePath
	return files.MemoryIntegrity(root, dirID)
}

// ReadMemoryFile returns one fact file's content for the editor's load.
// Read-only: no SSH gate.
func (s *MaintenanceService) ReadMemoryFile(dirID, fileName string) (string, error) {
	root := s.config.GetClaudeRootInfo().EffectivePath
	return files.ReadMemoryFile(root, dirID, fileName)
}

// WriteMemoryFile saves a fact file verbatim; the files layer refuses while the
// consolidation lock is present. SSH-gated + serialized under s.mu, matching
// every other mutating method on this service.
func (s *MaintenanceService) WriteMemoryFile(dirID, fileName, content string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.sshActive() {
		return errSSHActive
	}
	root := s.config.GetClaudeRootInfo().EffectivePath
	return files.WriteMemoryFile(root, dirID, fileName, []byte(content))
}

// ApplyMemoryIndexFix applies one byte-exact MEMORY.md index edit; the files
// layer re-derives the fix from a fresh integrity scan and refuses under the
// consolidation lock. SSH-gated + serialized under s.mu.
func (s *MaintenanceService) ApplyMemoryIndexFix(dirID string, fix files.MemoryIndexFix) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.sshActive() {
		return errSSHActive
	}
	root := s.config.GetClaudeRootInfo().EffectivePath
	return files.ApplyMemoryIndexFix(root, dirID, fix)
}

// DeleteMemoryFile trashes a fact file (user-authored — trash policy, never
// os.Remove). SSH-gated + serialized under s.mu; calls the PACKAGE-LEVEL
// maintenance.TrashItems (not s.TrashItems) to avoid re-entrant s.mu locking,
// and mutes the file watcher for the move — mirroring DeleteInstructionFile.
func (s *MaintenanceService) DeleteMemoryFile(dirID, fileName string) (maintenance.TrashReceipt, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.sshActive() {
		return maintenance.TrashReceipt{}, errSSHActive
	}

	root := s.config.GetClaudeRootInfo().EffectivePath
	dest, err := files.ResolveMemoryFilePath(root, dirID, fileName)
	if err != nil {
		return maintenance.TrashReceipt{}, err
	}

	roots, err := s.resolveRoots()
	if err != nil {
		return maintenance.TrashReceipt{}, err
	}
	appDataDir, err := config.AppDataDir()
	if err != nil {
		return maintenance.TrashReceipt{}, err
	}

	emitEvent("maintenance:mute-watcher", map[string]any{"muted": true})
	defer emitEvent("maintenance:mute-watcher", map[string]any{"muted": false})

	return maintenance.TrashItems(roots, appDataDir, []string{dest})
}

File: frontend/src/shared/types/api/memory.ts
// Memory manager DTOs (Week 28). Kind-prefixed dir IDs writes take instead of a
// path, per-dir integrity findings, and the byte-exact MEMORY.md index fixes
// the two auto-applicable finding kinds carry.

// MemoryDir is one addressable memory directory. id is a kind-prefixed,
// server-derived token ("project:<encoded>" / "agent:<name>"); label is the
// human-decoded name; path is absolute. kind ∈ "project" | "agent".
export interface MemoryDir {
  id: string;
  label: string;
  path: string;
  kind: string;
}

// MemoryFile is one fact file with its parsed frontmatter. fileName is the leaf
// on disk; name/description/type come from the frontmatter (type ∈
// user | feedback | project | reference).
export interface MemoryFile {
  fileName: string;
  name: string;
  description: string;
  type: string;
}

// MemoryIndexFix is a byte-exact MEMORY.md edit a finding proposes. op is "add"
// (append the line) or "remove" (drop the verbatim line). Passed back to
// applyMemoryIndexFix verbatim.
export interface MemoryIndexFix {
  op: string;
  line: string;
}

// MemoryFinding is one integrity issue. kind ∈ orphan-file | dangling-index |
// dangling-link | duplicate-slug. fix is non-null only for orphan-file (add)
// and dangling-index (remove); dangling-link and duplicate-slug are
// informational (fix null).
export interface MemoryFinding {
  kind: string;
  file: string;
  detail: string;
  fix: MemoryIndexFix | null;
}

export interface MemoryReport {
  dir: MemoryDir;
  files: MemoryFile[];
  findings: MemoryFinding[];
}
